src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newState = [];
+  let stateCopy = Object.assign({}, state);
+  for (let action in actions) {
+    const { type, extraData, keysToRemove } = actions[action];
+
+    if (action > 0) {
+      stateCopy = Object.assign({}, newState[action - 1]);
+    }
+
+    let currentItem = {};
+
+    switch (type) {
+      case 'addProperties':
+        currentItem = { ...stateCopy, ...extraData };
+        break;
+
+      case 'removeProperties':
+        for (let l = 0; l < keysToRemove.length; l++) {
+          delete stateCopy[keysToRemove[l]];
+        }
+        currentItem = { ...stateCopy };
+        break;
+
+      case 'clear':
+        break;
+    }
+    newState.push(currentItem);
```

Comment:
```suggestion
    switch (type) {
      case 'addProperties':
        stateCopy = { ...stateCopy, ...extraData };
        break;

      case 'removeProperties':
        for (let l = 0; l < keysToRemove.length; l++) {
          delete stateCopy[keysToRemove[l]];
        }
        break;

      case 'clear':
        stateCopy = {};
        break;
    }
    newState.push({...stateCopy});
```

Reply:
Ah, because `stateCopy` automaticaly has previous state in it since last itteration, we just have to perform operations directly without temp variables. So I essentially had duplicate variable. Thank you